<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-и-предобработка-данных" data-toc-modified-id="Обзор-и-предобработка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор и предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Устранение-пропусков" data-toc-modified-id="Устранение-пропусков-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Устранение пропусков</a></span></li><li><span><a href="#Изучение-аномалий" data-toc-modified-id="Изучение-аномалий-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Изучение аномалий</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Исследовательский-анализ-данных" data-toc-modified-id="Исследовательский-анализ-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследовательский анализ данных</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок" data-toc-modified-id="Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Есть ли зависимость между количеством детей и возвратом кредита в срок</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок</a></span></li></ul></li><li><span><a href="#Общий-вывод-и-рекомендации" data-toc-modified-id="Общий-вывод-и-рекомендации-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Общий вывод и рекомендации</a></span></li></ul></div>

# Исследование надежности заемщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Цель исследования** — оценить влияние отдельных характеристик клиентов банка на факт погашения кредита в срок.

**Ход исследования** построен в три этапа:

 1. Обзор данных и предобработка данных.
 3. Проверка гипотез о влиянии отдельных характеристик и факта погашения кредита в срок.
 4. Выводы и рекомендации для учета при построении модели кредитного скоринга.

## Обзор и предобработка данных

In [1]:
# Импортируем библиотеки
import pandas as pd

In [2]:
# Откроем файл с данными
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [4]:
# Функция первичного обзора данных
def review(df):
    display(df.head(5))
    print(df.info())
    print('Пропуски:', df.isna().sum())
    print('Явные дубликаты:')
    if df.duplicated().sum() > 0:
        print(df.duplicated().sum())
    else:
        print('Не найдено')

In [5]:
# Используем функцию для первичного обзора
review(data)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None
Пропуски: children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family

**Описание данных:**

-	children — количество детей в семье
-	days_employed — общий трудовой стаж в днях
-	dob_years — возраст клиента в годах
-	education — уровень образования клиента
-	education_id — идентификатор уровня образования
-	family_status — семейное положение
-	family_status_id — идентификатор семейного положения
-	gender — пол клиента
-	income_type — тип занятости
-	debt — имел ли задолженность по возврату кредитов
-	total_income — ежемесячный доход
-	purpose — цель получения кредита

Первый взгляд на датасет дает понять, что они, по всей видимости, были собраны из ответов пользователей. Данные достаточно "грязные", содержат пропуски, дубликаты. 

### Устранение пропусков

В датасете есть пропуски всего в двух столбцах - days_employed (общий трудовой стаж) и total_income (ежемесячный доход). Число пропущенных значений в первом и втором столбце равные, по 2174 в каждом. Также пропуски в этих слолбцах появляются одновременно. При этом портреты клиентов по остальным характеристикам (пол, возраст, образование, количество детей и др.) самые разнообразные.

Поэтому действительное отсутствие стажа / дохода, невнимательность при заполнении полей не могут рассматриваться как возможные причины появления пропусков.

**Итак, пропуски могли появиться по следующим причинам**:
- Ошибка программы при вводе данных
- Ошибка сервера при выгрузке данных

В случае необходимости заполнения пропусков для количественных переменных есть несколько решений, и все они имеют свои особенности:
- **Заполнение нулем** - может сильно исказить анализ, если проставить их безосновательно. Может применяться, только если действительно есть основания считать, что пропуски означают нулевой показатель. Например, если индивидуальный предприниматель не заполнил поле с количеством сотрудников, потому что у него их 0.
- **Заполнение средним** - также может быть необъективным, особенно если есть аномальные значения или если амплитуда данных слишком большая (_"У Пети 10 яблок, а у Васи —  0. В среднем у каждого по 5 яблок"_).
- **Заполнение медианой** - наиболее оптимальное решение, т.к. она более "справдлива", устойчива к аномальным значениям и "выбросам" данных. 

При заполнении медианным значением при возможности нужно отфильтровать категорию по наиболее влиятельным показателям. Так и было сделано в текущем проекте - пропуски в уровне дохода заполнены медианой по типу занятости.

На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца income_type. Например, у человека с типом занятости сотрудник пропуск в столбце total_income должен быть заполнен медианным доходом среди всех записей с тем же типом.

In [6]:
# Заполнение пустот в столбцах income_type и total_income медианным доходом по типу занятости
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Изучение аномалий

В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце days_employed. 

In [7]:
# Заменим все отрицательные значения положительными с помощью метода abs()
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведем медианное значение трудового стажа days_employed в днях.

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть. Тем более этот столбец не понадобится вам для исследования.

In [9]:
# Выведем перечень уникальных значений столбца children.
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

В столбце children есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения.

In [10]:
# Удалим строки с аномальными значениями
data = data[(data['children'] != -1) & (data['children'] != 20)]

Заполним пропуски в столбце c трудовым стажем медианными значениями по каждого типа занятости.

In [11]:
# Заполним пропуски в столбце days_employed медианными значениями по каждого типа занятости income_type 
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

### Изменение типов данных

Заменим тип данных в столбце total_income на целочисленный.

In [12]:
# Изменим тип данных в столбце total_income на int
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

Обработаем неявные дубликаты в столбце education. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру. 

In [13]:
# Приведем к нижнему регистру значения столбца education
data['education'] = data['education'].str.lower()

Посмотрим остались ли явные дубликаты после всех проведенных манипуляций.

In [15]:
# Количество строк-дубликатов в данных
data.duplicated().sum()

71

In [16]:
# Удаление строк-дубликатов
data = data.drop_duplicates()

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`.

In [17]:
# Функция категоризации доходов
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
# Используем функцию категоризации доходов
data['total_income_category'] = data['total_income'].apply(categorize_income)

Рассмотрим что указано в качестве уникальных целей взятия кредита.

In [19]:
# Перечень уникальных целей взятия кредита из столбца purpose
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.

In [20]:
# Функция категоризации целей взятия кредита
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
# Используем функцию категоризации целей взятия кредита
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследовательский анализ данных

### Есть ли зависимость между количеством детей и возвратом кредита в срок

In [22]:
# Сначала посмотрим как часто берут кредиты люди с разным количеством детей.
print(data['children'].value_counts())

0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64


In [23]:
# Проверим количество клиентов, имеющих задолженности по кредитам, методом группировки по количеству детей.
# Отсортируем данные по убыванию.
print(data.groupby('children')['debt'].sum().sort_values(ascending=False))

# Выведем долю должников о общего числа клиентов с определенным количеством детей через среднюю.
# Отсортируем данные по убыванию.
print(data.groupby('children')['debt'].mean().sort_values(ascending=False))

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64
children
4    0.097561
2    0.094542
1    0.092346
3    0.081818
0    0.075438
5    0.000000
Name: debt, dtype: float64


**Вывод:** Сначала может показаться, что должников больше всего среди основного клиентского сегмента банка - у людей без детей. Но если сравнить долю должников, то эта категория наоборот наиболее надежная _(клиентов с 5 детьми не считаем, их всего около 0,04% от общего числа)_. 
Зависимость между количеством детей и возвратом кредита в срок есть: **люди с детьми чаще имеют задолженности**. При этом точное количество детей может влиять по-разному - доля должников с 3 детьми меньше, чем доля должников с 2 детьми.

### Есть ли зависимость между семейным положением и возвратом кредита в срок

In [24]:
# Проверим зависимость тем же способом, как в предыдущем пункте.

# Сначала посмотрим как часто берут кредиты люди с разным семейным статусом.
print(data['family_status'].value_counts())

# Проверим количество клиентов, имеющих задолженности.
# Отсортируем данные по убыванию.
print(data.groupby('family_status')['debt'].sum().sort_values(ascending=False))

# Выведем долю должников через среднюю.
# Отсортируем данные по убыванию.
print(data.groupby('family_status')['debt'].mean().sort_values(ascending=False))

женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: family_status, dtype: int64
family_status
женат / замужем          927
гражданский брак         385
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
Name: debt, dtype: int64
family_status
Не женат / не замужем    0.097639
гражданский брак         0.093130
женат / замужем          0.075606
в разводе                0.070648
вдовец / вдова           0.066246
Name: debt, dtype: float64


**Вывод:** Зависимость между семейным положением и возвратом кредита в срок есть: **люди, не состоящие (и не состоявшие ранее) в официальном браке, чаще имеют задолженности**. Наиболее надежной категорией лиц являются вдовцы / вдовы, наименее надежной - не женатые / не замужние, а также состоящие в гражданском браке. 

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок

In [25]:
# Проверим зависимость тем же способом, как в предыдущих пунктах.

# Сначала посмотрим как часто берут кредиты люди с разным уровнем доходов. Обратимся к столбцу с категориями дохода A-E.
# - 0–30000 — `'E'`;
# - 30001–50000 — `'D'`;
# - 50001–200000 — `'C'`;
# - 200001–1000000 — `'B'`;
# - 1000001 и выше — `'A'`.
print(data['total_income_category'].value_counts())

# Проверим количество клиентов, имеющих задолженности.
# Отсортируем данные по убыванию.
print(data.groupby('total_income_category')['debt'].sum().sort_values(ascending=False))

# Выведем долю должников через среднюю.
# Отсортируем данные по убыванию.
print(data.groupby('total_income_category')['debt'].mean().sort_values(ascending=False))

C    15921
B     5014
D      349
A       25
E       22
Name: total_income_category, dtype: int64
total_income_category
C    1353
B     354
D      21
A       2
E       2
Name: debt, dtype: int64
total_income_category
E    0.090909
C    0.084982
A    0.080000
B    0.070602
D    0.060172
Name: debt, dtype: float64


**Вывод:**  Между уровнем дохода и возвратом кредита в срок **не наблюдается прямой зависимости**. Если убрать из списка людей с доходом A (1000001 и выше) и E (0–30000) люди, составляющих около 0,1% от общего числа, то наименее надежными являются клиенты с доходом C (50001–200000 руб.), далее - клиенты с доходом B (200001–1000000 руб.), а наиболее - клиенты с доходом D (30001–50000 руб.)


### Как разные цели кредита влияют на его возврат в срок

In [26]:
# Проверим зависимость тем же способом, как в предыдущих пунктах.

# Сначала посмотрим как часто берут кредиты люди с разными целями на кредит. Обратимся к столбцу с категориями.
print(data['purpose_category'].value_counts())

# Проверим количество клиентов, имеющих задолженности.
# Отсортируем данные по убыванию.
print(data.groupby('purpose_category')['debt'].sum().sort_values(ascending=False))

# Выведем долю должников через среднюю.
# Отсортируем данные по убыванию.
print(data.groupby('purpose_category')['debt'].mean().sort_values(ascending=False))

операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64
purpose_category
операции с недвижимостью    780
операции с автомобилем      400
получение образования       369
проведение свадьбы          183
Name: debt, dtype: int64
purpose_category
операции с автомобилем      0.093480
получение образования       0.092528
проведение свадьбы          0.079118
операции с недвижимостью    0.072551
Name: debt, dtype: float64


**Вывод:** **Наиболее ответственно** люди относятся к выплате самого популярного вида кредитов - **на операции с недвижимостью**. Задолженности чаще всего допускаются в случаях, когда кредит был взят на операции с автомобилем. 

## Общий вывод и рекомендации

Было проведено исследование влияния социальных характеристик клиента банка на возврат кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Предобработка данных**
- В данных были 10% пропусков в столбцах с информацией по доходу и трудовому стажу. Для исследования нужны были данные по доходу клиентов, поэтому здесь пропуски заполнены медианой по типу занятости. 
- В столбце с информацией о количестве детей и о трудовом стаже были случаи с аномальными значениями. 
- Для очистки данных от дубликатов было удалено 0,3% строк.


**Для исследования поставлены четыре гипотезы:**
1. Существует зависимость между количеством детей и возвратом кредита в срок
2. Существует зависимость между семейным положением и возвратом кредита в срок
3. Существует зависимость между уровнем дохода и возвратом кредита в срок
4. Разные цели кредита влияют на его возврат в срок

Исследование выявило следующие зависимости:

**Гипотеза 1.** Частично подтверждена.
Зависимость между количеством детей и возвратом кредита в срок есть: **люди с детьми чаще имеют задолженности, чем люди без детей**. То есть на возврат крелита в срок влияет именно факт наличия или отсутствия детей, при этом по имеющимся данным пока нельзя сделать вывод по влиянию точного количества детей. Для детального анализа нужна более репрезентативная выборка, чтобы исследуемые группы были близких размеров.

Количество детей | Доля должников
-------------    | -------------
4                |  0.097561
2                |  0.094542
1                |  0.092346
3                |  0.081818
нет детей        |  0.075438

**Гипотеза 2.** Подтверждена.
Зависимость между семейным положением и возвратом кредита в срок есть: **люди, не состоящие (и не состоявшие ранее) в официальном браке, чаще имеют задолженности**. Наименее надежной - не женатые / не замужние, а также состоящие в гражданском браке. Наиболее надежной категорией лиц являются вдовцы / вдовы, но их доля среди клиентов банка довольная маленькая.

Семейное положение     | Доля должников
-------------          | -------------
не женат / не замужем  |  0.097639
гражданский брак       |  0.093130
женат / замужем        |  0.075606
в разводе              |  0.070648
вдовец / вдова         |  0.066246

**Гипотеза 3.** Частично подтверждена. 
Между уровнем дохода и возвратом кредита в срок **не наблюдается прямой зависимости**. Если убрать из списка людей с доходом A (1000001 и выше) и E (0–30000) люди, составляющих около 0,1% от общего числа, то наименее надежными являются клиенты с доходом C (50001–200000 руб.), далее - клиенты с доходом B (200001–1000000 руб.), а наиболее - клиенты с доходом D (30001–50000 руб.). Для подробного анализа нужна более широкая выборка данных по клиентам с доходами 0–30000 руб., от 1000001 руб. и выше.
Также следует обратить внимание, что 10% данных были заполнены искусственно медианным значением.

Уровень дохода         | Доля должников
-------------          | -------------
C (50001–200000 руб.)  | 0.084982
B (200001–1000000 руб.)|0.070602
D (30001–50000 руб.)   |0.060172

**Гипотеза 4.** Подтверждена. 
**Наиболее ответственно** люди относятся к выплате самого популярного вида кредитов - **на операции с недвижимостью**. Задолженности чаще всего допускаются в случаях, когда кредит был взят на операции с автомобилем. 

Цель получения кредита  | Доля должников
-------------           | -------------
операции с автомобилем  |   0.093480
получение образования   |   0.092528
проведение свадьбы      |   0.079118
операции с недвижимостью|   0.072551


**Рекомендации**
1. При построении модели кредитного скоринга учесть результаты подтверждения 2 и 4 гипотезы (семейное положение и цели получения кредита)
2. Для проверки 1 и 3 гипотез нужны данные по недостающим группам.
3. Исследовать причину появления пропусков и дубликатов (вопрос к разработчикам, возможно была техническая ошибка при загрузке или выгрузке данных), аномалий (возможно это вопрос человеческого фактора при заполнении, может нужно ввести проверку вводимых данных, чтобы не было например отрицательных значений трудового стажа).
